### PR TITLE
Split providers out of the main "airflow/" tree into a UV workspace project

### LINKED ISSUE #42857: Re-organize repo into multiple projects using uv's workspaces "core" (scheduler + web/api servers), "providers" (a single project for all providers for now), "task-sdk" (new dist for this project)
List discussion https://lists.apache.org/thread/dyv5jhvt65xs6l5o2byc2b67f4wlwf6r

We will try out uv's workspace feature for this.

The projects we want are:

 - "core", containing the scheduler, web/api servers, dag parser etc.
 - "providers", a single project containing all the provivders
 - "task-sdk", the new dist that is key to this AIP.

There will be other top level folders such as the helm chart etc.

Tests should _probably_ be split to live alongside the projects.

Example code layout (picking a few representative files in each group):

```
core/pyproject.toml/
core/src/airflow/__init__.py
core/src/airflow/models/…
core/src/airflow/api_fastapi/…
core/src/airflow/ui/package.json # React UI etc.
core/tests/jobs/test_scheduler_job.py
providers/src/airflow/providers/celery/provider.yaml
providers/src/airflow/providers/celery/__init__.py
providers/tests/airflow/providers/celery/executors/test_celery_executor.py
providers/src/airflow/providers/cncf/kubernetes/provuder.yaml
providers/src/airflow/providers/cncf/kubernetes/__init__.py
task-sdk/pyproject.toml
task-sdk/src/airflow/sdk/__init__.py
task-sdk/src/airflow/sdk/defintions/dag.py
task-sdk/src/airflow/sdk/defintions/task_group.py
task-sdk/tests/airflow/sdk/defintions/test_task_group.py
```

### PR DESCRIPTION
**Pull request by ashb**
closes https://github.com/apache/airflow/issues/42857

As discussed https://lists.apache.org/thread/dyv5jhvt65xs6l5o2byc2b67f4wlwf6r, this is the first part of the new layout.

This is only a partial split so far. It moves all the code and tests, but the provider release generation code hasn't moved yet as it's a few weeks until the next batch, and I wan't to try and keep this already huge PR as small as possible. Likewise the creation of `core/` will be a follow on PR

In addition to the straight file rename the other changes I had to make here are:
- Some mypy/typing fixes.

  Mypy can be fragile about what it picks up when, so maybe some of those
  changes were caused by that. But the typing changes aren't large.

- Improve typing in common.sql type stub

  Again, likely a mypy file oddity, but the types should be safe

- Removed the `check-providers-init-file-missing` check

  This isn't needed now that airflow/providers shouldn't exist at all in the
  main tree.

- Create a "dev.tests_common" package that contains helper files and common
  pytest fixtures

  Since the provider tests are no longer under tests/ they don't automatically
  share the fixtures from the parent `tests/conftest.py` so they needed
  extracted.

  Ditto for `tests.test_utils` -- they can't be easily imported in provider
  tests anymore, so they are moved to a more explicit shared location.

In future we should switch how the CI image is built to make better use of UV caching than our own approach as that would remvoe a lot of custom code.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
Smalll comment (I am returning back from holidays) and slowly going through the two week's history.

> In future we should switch how the CI image is built to make better use of UV caching than our own approach as that would remvoe a lot of custom code.

I tried various things before (including `uv` caching and local cache mounts) and I could not find better solution. With all the other approaches I tried, approach is way more efficient for size and time and especially shines because with the current approach you have a single remote cache layer in ghcr.io that is used even if dependencies change, and you do not have cache duplication.

There are two problems with using uv cache:

1) Enabling UV cache (I tested it having similar hope when UV was released) makes our image 4 GB instead of 2GB (can't remember exactly but it was almost 2x factor). also it does not work with docker layer caching. It's very time-efficient to keep the UV cache but yoy pay a huge price of used space. Note that UV will cache not only the actual packages that it downloads but also the packages it considers as candidates - which in case of significant backtracking might mean that you keep multiple versions of the same package in the cache - this adds up very quickly and uv cache is really huge.

UV cache benefits from the fact that the same cache is potentially reused by multiple projects you have installed on the same machine. The UV cache is stored in  HOME dir and reused between all venvs used in the same machine - so in docker it's nearly useless because it multiplies the space used for both - installed and cached files but cached files are not reusable in any way for other venvs (because we have only one venv in docker container). And since layers in docker image cannot be removed, even if we delete the cache after final installation step - the cache will remain in the docker image layer additionally to installed packages - increasing significantly size of the final image.

Of course - you could potentially use locally mounted cache volumes in docker to keep the UV cache on,  in order to not pollute docke image, but that would only be useful after you build your image locally once and you won't be able to use remotely cached layers which are currently used when you build the image. 

Using remote cache in the way we do it now, is much nicer, because the cache is actually the same as most installed packages (with the exception of the new/ updated packages change after the original cached layer is created.  This has the nice property that every time new patchlevel base image is released by python (every two-three weeks or so) the cache is rebuit from scratch, and we get a new "fresh" main cache with latest package versions. 

2) Another problem is that you will not avoid the problem that you need to have all the "source" dependency files copied to the image layer in order to run `uv pip install` command.  

Whatever that command will produce as cache will be invalidated the moment the source dependency files (pyproject.toml, hatch_build.py, provider.yaml files, generated/provider_dependencies.json) - so if you want to effectively use cache, you need to build it BEFORE you copy any of those files into the new image. This is precisely what "our" caching is doing - it does not use any "local file" and does not copy them to the image, but uses "https://github" URL to install the dependencies for the first time - which means that next time that layer will not be invalidated when pyproject.toml or hatch_build.py or any other files that decide about dependencies will change.

Now - I'd love to make it simpler, but so far all the attempts to find a better solution failed because the one we have is the only one that has good "caching" properties.

**Comment by potiuk** [comment]
But maybe you can come up with a better approach that I have not thought about and tried already, of course :)

**Comment by ashb** [comment]
Yeah, that comment was mostly a note to myself, and not something I plan on even looking at as part of this work.

**Comment by ashb** [comment]
I think I'm getting closer now 🤞🏻 .

Oddly enough a lot of the time the er diagram static check is failing in CI , even though I've run `breeze static-checks -a update-er-diagram` and it says pass.

**Comment by ashb** [comment]
Something is failing with the kube tests, it's not picking up the providers for running tests, however running the same command locally passes 🤔

File: providers/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+[build-system]
+requires = ["hatchling", "hatch-vcs"]
+build-backend = "hatchling.build"
+
+[project]
+name = "local-providers"
+version = "0.1.0"
+description = "Placeholder package for local/from-sources providers."
+requires-python = ">=3.8, <3.13"
```
Comment (ashb) [inline providers/pyproject.toml:26]: This purposefully stays at 3.8 in this PR, as if I change it to 3.9 then there are a load of upgrade diffs applied and I don't want to mix up this already massive PR with more changes than are strictly required.

Once it lands I will change this and apply all the upgrades it results in.

**Review by kaxil — APPROVED** [review]

**Review by phanikumv — APPROVED** [review]

**Comment by ashb** [comment]
3030 files are renamed without changes:

```
git di HEAD origin/main --stat | grep -F '|    0' | wc -l
    3030
```

**Review by ephraimbuddy — APPROVED** [review]

**Review by jscheffl — APPROVED** [review]
I inspected the 55k LoC from GIT DIFF yesterday and that looked good. LGTM!
(Cherry-picking to v2-10 after this structure change will be a nightmare though :-D )

**Comment by potiuk** [comment]
> (Cherry-picking to v2-10 after this structure change will be a nightmare though :-D )

Not really :D - we are not cherry-picking providers. And git actually is able to follow renames when cherry-picking.

**Comment by ashb** [comment]
For anyone coming to this PR after the fact, yes we know there are still a few rough edges as a result of this.

Currently known issues:

- [ ] Pycharm doesn't like resolving the imports of `airflow.providers.x` (Something to do with the content of the main `airflow/__init__.py` that means it doesn't think it's a namespace package?)
- [ ] `breeze testing tests provider/tests/x.py` currently fails. (Temp workaround: `breeze shell` then run `pytest provider/tests/x.py` directly. Bug is that the `testing test` command includes `tests` in the pytest command even when another file is specified, leading to pytest loading tests/conftest.py and providers/tests/conftest.py which it doesn't like as they are both "top level")

**Comment by dstandish** [comment]
#protm

**Comment by potiuk** [comment]
> breeze testing tests provider/tests/x.py currently fails. (Temp workaround: breeze shell then run pytest provider/tests/x.py directly. Bug is that the testing test command includes tests in the pytest command even when another file is specified, leading to pytest loading tests/conftest.py and providers/tests/conftest.py which it doesn't like as they are both "top level")

This was never supposed to work. the parameters that you can pass to "testing tests" is really "extra pytest paramaters" you can pass - and it's not really inteded to use "pytest files" there. That's why for example the extra parameters are not autocomplete'able with files/paths. The "tests" command is really meant to run the whole "test type" and you can easily use it now. 

For example running all airbyte tests can be achieved by:

```bash
breeze testing tests --test-type "Providers[airbyte]"
```

And it nicely works now.

**Comment by potiuk** [comment]
(so I think indeed all known "teething" problems" are solved - and i can try to attempt to split out individual providers).

**Pull request merged**